File: pip-install-series/parsing/ecs_v1/ecs.cpp
#include <format>
#include <map>
#include <stdexcept>
#include <string>
#include <unordered_set>
#include <vector>
#include "validate.hpp"

// Entity
using Guest = int;

// Component
using Name = std::string;

// World
class RestaurantBar {
public:
    std::map<Guest, Name> names;
    std::unordered_set<Guest> drinkers;

    void parse_guest(const std::string& name, const std::string& raw_age) {
        int age = std::stoi(raw_age);
        if (!is_valid(age, 18)) {
            throw std::invalid_argument(std::format("Guest {} is underage: {}", name, age));
        }
        Guest guest = admit(name);
        if (is_valid(age, 21)) {
            drinkers.insert(guest);
        }
    }

private:
    int id = 0;

    Guest admit(const std::string& name) {
        Guest guest = id++;
        names[guest] = name;
        return guest;
    }
};

// System
std::vector<std::string> serve_food(const RestaurantBar& world) {
    std::vector<std::string> orders;
    for (const auto& [guest, name] : world.names) {
        orders.push_back(std::format("Ribeye for {}", name));
    }
    return orders;
}

std::vector<std::string> serve_drink(const RestaurantBar& world) {
    std::vector<std::string> orders;
    for (const Guest& guest : world.drinkers) {
        orders.push_back(std::format("Wine for {}", world.names.at(guest)));
    }
    return orders;
}

File: pip-install-series/parsing/ecs_v1/ecs.py
from collections.abc import Iterator
from typing import Any, NewType
from validate import is_valid

# Entity
Guest = NewType("Guest", int)

# Component
Name = NewType("Name", str)

# World
class RestaurantBar:
    def __init__(self):
        self._id: int = 0
        self.names: dict[Guest, Name] = {}
        self.drinkers: set[Guest] = set()

    def _admit(self, name: str) -> Guest:
        guest = Guest(self._id)
        self._id += 1
        self.names[guest] = Name(name)
        return guest

    def parse_guest(self, raw_name: Any, raw_age: Any) -> None:
        name, age = str(raw_name), int(raw_age)
        if not is_valid(age, 18):
            raise ValueError(f"Guest {name} is underage: {age}")
        guest = self._admit(name)
        if is_valid(age, 21):
            self.drinkers.add(guest)

# System
def serve_food(world: RestaurantBar) -> Iterator[str]:
    return (f"Ribeye for {name}" for name in world.names.values())

def serve_drink(world: RestaurantBar) -> Iterator[str]:
    return (f"Wine for {world.names[guest]}" for guest in world.drinkers)
